### ISSUE TITLE
TPC-H Q1 lead to TiKV Panic

### ISSUE THREAD

**Issue by breeswish** [tikv/tikv#5404]
## Bug Report

When TPC-H Q1 Decimal is modified to Double, TiKV will panic with latest TiDB. This means that our type check is not comprehensive and there are some missing part that illegal request may crash TiKV.

```
[2019/09/04 23:05:34.948 +08:00] [FATAL] [lib.rs:489] ["Cannot call `$ name` over to a Int column"] [backtrace="stack backtrace:\n   0:     0x55e52d561efd - backtrace::backtrace::libunwind::trace::h52557d88869
71e0e\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54\n                         - backtrace::backtrace::trace::h16a90011d
c166192\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70\n   1:     0x55e52d558c35 - tikv_util::set_panic_hook::{{closure}}::h2b
00fe479cdfa106\n                        at /home/pingcap/wenxuan/tikv/target/release/build/backtrace-96a937f63df81ff4/out/capture.rs:79\n   2:     0x55e52d711eac - std::panicking::rust_panic_with_hook::h3ccc79
b44ba3c353\n                        at src/libstd/panicking.rs:481\n   3:     0x55e52d711c8d - std::panicking::continue_panic_fmt::h6da01b9bb2c73c79\n                        at src/libstd/panicking.rs:384\n
4:     0x55e52d711c3b - std::panicking::begin_panic_fmt::ha4fbbae8c2cc86c2\n                        at src/libstd/panicking.rs:339\n   5:     0x55e52d315f38 - tidb_query::codec::data_type::vector::VectorValue:
:push_real::h5a4107af03045e57\n                        at components/tidb_query/src/codec/data_type/vector.rs:408\n                         - <tidb_query::codec::data_type::vector::VectorValue as tidb_query::c
odec::data_type::vector::VectorValueExt<ordered_float::NotNan<f64>>>::push::h43e824f1ae7eb8c9\n                        at components/tidb_query/src/codec/data_type/vector.rs:422\n   6:     0x55e52d31ff8c - <ti
db_query::aggr_fn::impl_sum::AggrFnStateSum<T> as tidb_query::aggr_fn::ConcreteAggrFunctionState>::push_result::h31ccb1b13e171d69\n                        at components/tidb_query/src/aggr_fn/impl_sum.rs:0\n
                       - <F as tidb_query::aggr_fn::AggrFunctionState>::push_result::h7e53e67fc25da054\n                        at components/tidb_query/src/aggr_fn/mod.rs:200\n   7:     0x55e52cb5ce72 - tidb_
query::batch::executors::util::aggr_executor::AggregationExecutor<Src,I>::aggregate_partial_results::{{closure}}::h993dcd49281cd2e6\n                        at /home/pingcap/wenxuan/tikv/components/tidb_query/
src/batch/executors/util/aggr_executor.rs:245\n                         - <tidb_query::batch::executors::slow_hash_aggr_executor::SlowHashAggregationImpl as tidb_query::batch::executors::util::aggr_executor::A
ggregationExecutorImpl<Src>>::iterate_available_groups::hebabe1775b935169\n                        at /home/pingcap/wenxuan/tikv/components/tidb_query/src/batch/executors/slow_hash_aggr_executor.rs:323\n
                   - tidb_query::batch::executors::util::aggr_executor::AggregationExecutor<Src,I>::aggregate_partial_results::h2c52cd469bd2bcb7\n                        at /home/pingcap/wenxuan/tikv/component
s/tidb_query/src/batch/executors/util/aggr_executor.rs:233\n                         - tidb_query::batch::executors::util::aggr_executor::AggregationExecutor<Src,I>::handle_next_batch::hacc46bb6d1fce08d\n
                   at /home/pingcap/wenxuan/tikv/components/tidb_query/src/batch/executors/util/aggr_executor.rs:215\n                         - <tidb_query::batch::executors::util::aggr_executor::AggregationE
xecutor<Src,I> as tidb_query::batch::interface::BatchExecutor>::next_batch::h9bb224c92880e6cb\n                        at /home/pingcap/wenxuan/tikv/components/tidb_query/src/batch/executors/util/aggr_executor
.rs:283\n                         - <tidb_query::batch::executors::slow_hash_aggr_executor::BatchSlowHashAggregationExecutor<Src> as tidb_query::batch::interface::BatchExecutor>::next_batch::h42eef9b9bbfe21ab\
n                        at /home/pingcap/wenxuan/tikv/components/tidb_query/src/batch/executors/slow_hash_aggr_executor.rs:42\n   8:     0x55e52cb76ef8 - <tidb_query::execute_stats::WithSummaryCollector<C,T>
as tidb_query::batch::interface::BatchExecutor>::next_batch::h395a7c58b23edb87\n                        at /home/pingcap/wenxuan/tikv/components/tidb_query/src/batch/interface.rs:97\n   9:     0x55e52cd0846a -
 <alloc::boxed::Box<T> as tidb_query::batch::interface::BatchExecutor>::next_batch::h1ba90e2b300034b2\n                        at /home/pingcap/wenxuan/tikv/components/tidb_query/src/batch/interface.rs:74\n
                      - tidb_query::batch::runner::BatchExecutorsRunner<SS>::handle_request::h01951cb6d9d3411d\n                        at /home/pingcap/wenxuan/tikv/components/tidb_query/src/batch/runner.rs:3
61\n                         - <tikv::coprocessor::dag::BatchDAGHandler as tikv::coprocessor::RequestHandler>::handle_request::h78b498544f9b2cd4\n                        at src/coprocessor/dag/mod.rs:108\n  10
:     0x55e52cb20cc0 - tikv::coprocessor::endpoint::Endpoint<E>::handle_unary_request_impl::{{closure}}::h01163cc4f56aaa8e\n                        at /home/pingcap/wenxuan/tikv/src/coprocessor/endpoint.rs:256
\n                         - <futures::future::and_then::AndThen<A,B,F> as futures::future::Future>::poll::{{closure}}::{{closure}}::h7faf20842b9ee5ff\n                        at /home/pingcap/.cargo/registry/
src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/and_then.rs:34\n                         - core::result::Result<T,E>::map::hfc7a0bc2b377775a\n                        at /rustc/311376d30dc1cfa622142a9
f50317b1e0cb4608a/src/libcore/result.rs:512\n                         - <futures::future::and_then::AndThen<A,B,F> as futures::future::Future>::poll::{{closure}}::hf1f76c39a364df54\n                        at
/home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/and_then.rs:33\n                         - futures::future::chain::Chain<A,B,C>::poll::h2ef3cc791adc3335\n
         at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/chain.rs:39\n                         - <futures::future::and_then::AndThen<A,B,F> as futures::future::Future
>::poll::hc77750e783739614\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/and_then.rs:32\n                         - futures::future::chain:
:Chain<A,B,C>::poll::hbf13d920837623fd\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/chain.rs:26\n  11:     0x55e52cb1dc63 - <futures::futu
re::then::Then<A,B,F> as futures::future::Future>::poll::h61399becb1d8ad57\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/then.rs:32\n
                   - <futures::future::lazy::Lazy<F,R> as futures::future::Future>::poll::h69055be73b984334\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1
.23/src/future/lazy.rs:82\n                         - <futures::sync::oneshot::Execute<F> as futures::future::Future>::poll::hcfeef5325b436d03\n                        at /home/pingcap/.cargo/registry/src/gith
ub.com-1ecc6299db9ec823/futures-0.1.23/src/sync/oneshot.rs:554\n  12:     0x55e52d5bee95 - <alloc::boxed::Box<F> as futures::future::Future>::poll::h65b8ae81556157ac\n                        at /home/pingcap/.
cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/mod.rs:113\n                         - futures::task_impl::Spawn<T>::poll_future_notify::{{closure}}::h6b5a9a9103d09b3c\n
         at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/mod.rs:289\n                         - futures::task_impl::Spawn<T>::enter::{{closure}}::h78df4f60652f0c15
\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/mod.rs:363\n                         - futures::task_impl::std::set::h9df7792d1b3f31b0\n
                       at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/std/mod.rs:78\n                         - futures::task_impl::Spawn<T>::enter::h434112265d71
7219\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/mod.rs:363\n                         - futures::task_impl::Spawn<T>::poll_future_noti
fy::hd3a109fd85007d16\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/mod.rs:289\n                         - tokio_threadpool::task::Task:
:run::{{closure}}::haeaa8b274b17996c\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.13/src/task/mod.rs:145\n                         - core::ops:
:function::FnOnce::call_once::h0d769f3f9db8a49d\n                        at /rustc/311376d30dc1cfa622142a9f50317b1e0cb4608a/src/libcore/ops/function.rs:231\n                         - <std::panic::AssertUnwind
Safe<F> as core::ops::function::FnOnce<()>>::call_once::hfd04c347a69f70e7\n                        at /rustc/311376d30dc1cfa622142a9f50317b1e0cb4608a/src/libstd/panic.rs:315\n                         - std::pa
nicking::try::do_call::h8c824e7ba56f5e09\n                        at /rustc/311376d30dc1cfa622142a9f50317b1e0cb4608a/src/libstd/panicking.rs:296\n                         - std::panicking::try::h3df016f4090736
46\n                        at /rustc/311376d30dc1cfa622142a9f50317b1e0cb4608a//src/libpanic_abort/lib.rs:29\n                         - std::panic::catch_unwind::h88c9ebeeb8e6a93a\n                        at
/rustc/311376d30dc1cfa622142a9f50317b1e0cb4608a/src/libstd/panic.rs:394\n                         - tokio_threadpool::task::Task::run::h456b9b5c1efe1a35\n                        at /home/pingcap/.cargo/registr
y/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.13/src/task/mod.rs:130\n                         - tokio_threadpool::worker::Worker::run_task2::h139ced29ff2f3eb8\n                        at /home/pingca
p/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.13/src/worker/mod.rs:567\n                         - tokio_threadpool::worker::Worker::run_task::h908cd04295a8b9bf\n
  at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.13/src/worker/mod.rs:459\n  13:     0x55e52d5bc91c - tokio_threadpool::worker::Worker::try_run_owned_task::ha335dce02310d
3f1\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.13/src/worker/mod.rs:390\n                         - tokio_threadpool::worker::Worker::try_run
_task::h837caba1219a4f2f\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.13/src/worker/mod.rs:297\n                         - tokio_threadpool::wo
rker::Worker::run::h770cc1bc86b7692c\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.13/src/worker/mod.rs:241\n  14:     0x55e52d5bc237 - tokio_th
readpool::worker::Worker::do_run::{{closure}}::{{closure}}::hd5f2089c62043e50\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.13/src/worker/mod.rs
:129\n                         - tokio_executor::global::with_default::{{closure}}::h10bd7d7d1ce36300\n                        at /home/pingcap/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.
1.7/src/global.rs:209\n                         - std::thread::local::LocalKey<T>::try_with::hc5899b8a56aa3e8d\n                        at /rustc/311376d30dc1cfa622142a9f50317b1e0cb4608a/src/libstd/thread/loca
l.rs:257\n                         - std::thread::local::LocalKey<T>::with::hb39515a27d8034e9\n                        at /rustc/311376d30dc1cfa622142a9f50317b1e0cb4608a/src/libstd/thread/local.rs:234\n
```

**Comment by breeswish** [comment]
TiKV received this kind of illegal request using latest TiDB:

```
[2019/09/05 01:12:34.166 +09:00] [WARN] [impl_sum.rs:34] ["aggr_def = tp: Sum children {tp: ColumnRef val: \"\\200\\000\\000\\000\\000\\000\\000\\001\" sig: CastIntAsInt field_type {tp: 5 flag: 4097 flen: 22 decimal: -1 collate: 63 charset: \"binary\"}} sig: CastIntAsInt field_type {tp: 8 flag: 128 flen: 21 decimal: -1 collate: 63 charset: \"binary\"}"]
[2019/09/05 01:12:34.166 +09:00] [WARN] [impl_sum.rs:34] ["aggr_def = tp: Sum children {tp: ColumnRef val: \"\\200\\000\\000\\000\\000\\000\\000\\001\" sig: CastIntAsInt field_type {tp: 5 flag: 4097 flen: 22 decimal: -1 collate: 63 charset: \"binary\"}} sig: CastIntAsInt field_type {tp: 8 flag: 128 flen: 21 decimal: -1 collate: 63 charset: \"binary\"}"]
```

@zz-jason

**Comment by breeswish** [comment]
If we use the original TPC-H schema, TiKV will also receive illegal request:

```
[2019/09/05 01:26:26.689 +09:00] [WARN] [impl_sum.rs:34] ["aggr_def = tp: Sum children {tp: ColumnRef val: \"\\200\\000\\000\\000\\000\\000\\000\\000\" sig: CastIntAsInt field_type {tp: 246 flag: 4097 flen: 15 decimal: 2 collate: 63 charset: \"binary\"}} sig: CastIntAsInt field_type {tp: 246 flag: 128 flen: 65 decimal: 6 collate: 63 charset: \"binary\"}"]
[2019/09/05 01:26:26.689 +09:00] [WARN] [impl_sum.rs:34] ["aggr_def = tp: Sum children {tp: ColumnRef val: \"\\200\\000\\000\\000\\000\\000\\000\\001\" sig: CastIntAsInt field_type {tp: 246 flag: 4097 flen: 15 decimal: 2 collate: 63 charset: \"binary\"}} sig: CastIntAsInt field_type {tp: 8 flag: 128 flen: 21 decimal: -1 collate: 63 charset: \"binary\"}"]
[2019/09/05 01:26:26.689 +09:00] [WARN] [impl_sum.rs:34] ["aggr_def = tp: Sum children {tp: ColumnRef val: \"\\200\\000\\000\\000\\000\\000\\000\\000\" sig: CastIntAsInt field_type {tp: 246 flag: 4097 flen: 15 decimal: 2 collate: 63 charset: \"binary\"}} sig: CastIntAsInt field_type {tp: 246 flag: 128 flen: 65 decimal: 6 collate: 63 charset: \"binary\"}"]
[2019/09/05 01:26:26.689 +09:00] [WARN] [impl_sum.rs:34] ["aggr_def = tp: Sum children {tp: ColumnRef val: \"\\200\\000\\000\\000\\000\\000\\000\\002\" sig: CastIntAsInt field_type {tp: 246 flag: 4097 flen: 15 decimal: 2 collate: 63 charset: \"binary\"}} sig: CastIntAsInt field_type {tp: 246 flag: 128 flen: 65 decimal: 6 collate: 63 charset: \"binary\"}"]
[2019/09/05 01:26:26.690 +09:00] [WARN] [impl_sum.rs:34] ["aggr_def = tp: Sum children {tp: ColumnRef val: \"\\200\\000\\000\\000\\000\\000\\000\\001\" sig: CastIntAsInt field_type {tp: 246 flag: 4097 flen: 15 decimal: 2 collate: 63 charset: \"binary\"}} sig: CastIntAsInt field_type {tp: 8 flag: 128 flen: 21 decimal: -1 collate: 63 charset: \"binary\"}"]
[2019/09/05 01:26:26.690 +09:00] [WARN] [impl_sum.rs:34] ["aggr_def = tp: Sum children {tp: ColumnRef val: \"\\200\\000\\000\\000\\000\\000\\000\\002\" sig: CastIntAsInt field_type {tp: 246 flag: 4097 flen: 15 decimal: 2 collate: 63 charset: \"binary\"}} sig: CastIntAsInt field_type {tp: 246 flag: 128 flen: 65 decimal: 6 collate: 63 charset: \"binary\"}"]
```

Which is, `SumLong(DecimalColumn)`.

**Comment by breeswish** [comment]
Fixed by #5429
